### ISSUE TITLE
NLLLoss reduce=True returning nan in float16

### ISSUE THREAD

**Issue by ychnh** [pytorch/pytorch#29429]
## 🐛 Bug
torch 1.3.1
torchvision 0.4.1
NLLLoss reduce=True doesn't seem to work in float16.
Also, training a model with loss1 in float16 doesn't seem to decrease the loss. (The model trains fine with loss1 in float32)

Code to reproduce
```
input = torch.rand(32,11,256,256).cuda().half()
target =torch.LongTensor(32,256,256).random_(0, 10).cuda()
loss1 = nn.NLLLoss(reduce=False).cuda().half()
loss2 = nn.NLLLoss(reduce=True, reduction='mean').cuda().half()
print(torch.mean(loss1(input,target)) )
print(loss2(input,target))
```

```
tensor(-0.4995, device='cuda:0', dtype=torch.float16)
tensor(nan, device='cuda:0', dtype=torch.float16)
```

**Comment by soumith** [comment]
i think we need to tease out how much of this is due to the extremely limited precision and dynamic range of float16, versus there being an actual bug.

**Comment by ychnh** [comment]
That is a good direction. I can look into it once I have time.

The following might give some insight to this issue.
I was able to get the model to work by using the Nvidia's APEX Automatic Mixed Precision.
It **did not** raise any issues like `Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 32768.0`

**Comment by soumith** [comment]
APEX AMP afaik will only do certain operations in fp16, and NLLLoss is not one of them.
